wagtail/contrib/settings/views.py

```diff
@@ -82,118 +86,280 @@ def redirect_to_relevant_instance(request, app_name, model_name):
+            instance = self.model.for_site(site)
 
-    elif issubclass(model, BaseGenericSetting):
-        queryset = model.base_queryset()
+            # Show a site switcher form if there are multiple sites
+            if Site.objects.count() > 1:
+                site_switcher = SiteSwitchForm(site, self.model)
+        else:
+            instance = get_object_or_404(self.model, pk=self.kwargs["pk"])
+
+        form = self.get_form()
+
+        context.update(
+            {
+                "setting_type_name": self.model._meta.verbose_name,
+                "instance": instance,
+                "edit_handler": get_setting_edit_handler(self.model).get_bound_panel(
+                    instance=instance, request=self.request, form=form
+                ),
+                "site": site,
+                "site_switcher": site_switcher,
+                "tabbed": isinstance(
+                    get_setting_edit_handler(self.model), TabbedInterface
+                ),
+            }
+        )
 
-        # Create the instance if we haven't already.
-        if queryset.count() == 0:
-            model.objects.create()
+        return context
 
-        instance = get_object_or_404(model, pk=pk)
-        form_id = instance.pk
+    def get_success_url(self):
+        site = None
+        if issubclass(self.model, BaseSiteSetting):
+            site = get_object_or_404(Site, pk=self.kwargs["pk"])
 
-        if request.method == "POST":
-            form = form_class(
-                request.POST, request.FILES, instance=instance, for_user=request.user
+        if site is not None:
+            return redirect(
+                "wagtailsettings:edit",
+                app_name=self.model._meta.app_label,
+                model_name=self.model._meta.model_name,
+                pk=site.pk,
             )
-
-            if form.is_valid():
-                with transaction.atomic():
-                    form.save()
-                    log(instance, "wagtail.edit")
-
-                messages.success(
-                    request,
-                    _("%(setting_type)s updated.")
-                    % {
-                        "setting_type": capfirst(setting_type_name),
-                        "instance": instance,
-                    },
-                )
-                return redirect("wagtailsettings:edit", app_name, model_name)
-            else:
-                messages.validation_error(
-                    request, _("The setting could not be saved due to errors."), form
-                )
         else:
-            form = form_class(instance=instance, for_user=request.user)
-
-        edit_handler = edit_handler.get_bound_panel(
-            instance=instance, request=request, form=form
-        )
+            return redirect(
+                "wagtailsettings:edit",
+                app_name=self.model._meta.app_label,
+                model_name=self.model._meta.model_name,
+                pk=self.object.pk,
```

Comment:
I think it should be safe to eliminate the whole logic and just `return self.get_edit_url()`?